core/connection/manager.go

```diff
@@ -732,6 +739,48 @@ func (manager *connectionManager) getCurrentSession() SessionInfo {
 	return manager.sessionInfo
 }
 
+func (manager *connectionManager) keepAliveLoop(channel p2p.Channel) {
+	// TODO: Remove this check once all provider migrates to p2p.
+	if channel == nil {
+		return
+	}
+
+	// Register handler for handling p2p keep alive pings from provider.
+	channel.Handle(p2p.TopicKeepAlive, func(c p2p.Context) error {
+		var ping pb.P2PKeepAlivePing
+		if err := c.Request().UnmarshalProto(&ping); err != nil {
+			return err
+		}
+
+		log.Debug().Msgf("Received p2p keepalive ping: %s", ping.Msg)
+		return c.OK()
+	})
+
+	// Send pings to provider.
+	var errCount int
+	for {
+		select {
+		case <-manager.ctx.Done():
+			return
+		case <-time.After(manager.params.KeepAlive.SendInterval):
+			sess := manager.getCurrentSession()
+			msg := &pb.P2PKeepAlivePing{
+				Msg: fmt.Sprintf("consumser SessionID: %s", sess.SessionID),
+			}
+			_, err := channel.Send(p2p.TopicKeepAlive, p2p.ProtoMessage(msg))
+			if err != nil {
+				log.Err(err).Msgf("Failed to send p2p keepalive ping. SessionID: %s", sess.SessionID)
+				errCount++
+				if errCount == manager.params.KeepAlive.MaxSendErrCount {
+					log.Error().Msgf("Max p2p keepalive err count reached, disconnecting. SessionID: %s", sess.SessionID)
+					manager.Disconnect()
+					return
+				}
+			}
```

Comment:
If we did a successful send, we can reset the error counter, otherwise, we will accumulate single errors over time. 
```suggestion
			} else {
				errCount = 0
			}
```

Reply:
Good point, thanks 👍